<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!mamba install html5lib==1.1 -y
!pip install lxml==4.6.4
!pip install nbformat>=4.2.0
!pip install pandas==1.3.3
!pip install requests==2.26.0
!pip install plotly==5.3.1

/bin/bash: line 1: mamba: command not found
/bin/bash: line 1: mamba: command not found


In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from IPython.display import display

/usr/local/lib/python3.10/dist-packages/requests/__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.16) or chardet (5.2.0)/charset_normalizer (2.0.12) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({})/charset_normalizer ({}) doesn't match a supported "


In [3]:
!pip install nbformat>=4.2.0

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [40]:
#I changed the 'make_graph' function using the 'display' method instead of `fig.show()` because that works better in Colab
#The original 'make_graph' function works ok in Jupyter Lab

def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing=.3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']

    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)

    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)

    fig.update_layout(showlegend=False, height=900, title=stock, xaxis_rangeslider_visible=True)

    # Display the figure using IPython's display method
    display(fig)

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [5]:
tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [6]:
tesla_data = tesla.history(period="max")

In [7]:
type(tesla_data)

pandas.core.frame.DataFrame

In [8]:
tesla_data.head()

Open      High       Low     Close     Volume  Dividends  \
Date                                                                       
2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

            Stock Splits  
Date                      
2010-06-29           0.0  
2010-06-30           0.0  
2010-07-01           0.0  
2010-07-02           0.0  
2010-07-06           0.0

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [9]:
tesla_data.reset_index(inplace=True)

In [10]:
tesla_data.head()

Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

In [11]:
tesla_data.shape

(3335, 8)

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.

In [12]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data  = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [13]:
soup = BeautifulSoup(html_data, 'html5lib')

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [14]:
tesla_revenue = pd.DataFrame(columns=["Date", "Revenue"])

for row in soup.find("tbody").find_all('tr'):
    col = row.find_all("td")
    date = col[0].text
    revenue = col[1].text
    tesla_revenue = tesla_revenue.append({"Date":date, "Revenue":revenue}, ignore_index=True)

In [15]:
tesla_revenue

Date  Revenue
0   2021  $53,823
1   2020  $31,536
2   2019  $24,578
3   2018  $21,461
4   2017  $11,759
5   2016   $7,000
6   2015   $4,046
7   2014   $3,198
8   2013   $2,013
9   2012     $413
10  2011     $204
11  2010     $117
12  2009     $112

Execute the following line to remove the comma and dollar sign from the `Revenue` column.


In [16]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

<ipython-input-16-2aef5327de36>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")


In [17]:
tesla_revenue

Date Revenue
0   2021   53823
1   2020   31536
2   2019   24578
3   2018   21461
4   2017   11759
5   2016    7000
6   2015    4046
7   2014    3198
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112

Execute the following lines to remove an null or empty strings in the Revenue column.


In [18]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [19]:
tesla_revenue.tail()

Date Revenue
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112

In [20]:
tesla_revenue.dtypes

Date       object
Revenue    object
dtype: object

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [21]:
gamestop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [22]:
gme_data = gamestop.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [23]:
gme_data.reset_index(inplace=True)

In [24]:
gme_data.head()

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620128  1.693349  1.603295  1.691666  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683251  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658001  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [25]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
#I won't need this if I use the function 'read_html' from pandas
html_data  = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [26]:
#I won't need this if I use the function 'read_html' from pandas
soup = BeautifulSoup(html_data, 'html5lib')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [27]:
read_html_pandas_data = pd.read_html(url)
read_html_pandas_data

[    GameStop Annual Revenue(Millions of US $)  \
 0                                        2020   
 1                                        2019   
 2                                        2018   
 3                                        2017   
 4                                        2016   
 5                                        2015   
 6                                        2014   
 7                                        2013   
 8                                        2012   
 9                                        2011   
 10                                       2010   
 11                                       2009   
 12                                       2008   
 13                                       2007   
 14                                       2006   
 15                                       2005   
 
    GameStop Annual Revenue(Millions of US $).1  
 0                                       $6,466  
 1                                       $8,285 

In [28]:
gme_revenue = read_html_pandas_data[0]

In [29]:
gme_revenue.columns = ("Date", "Revenue")

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [30]:
gme_revenue.tail()

Date Revenue
11  2009  $8,806
12  2008  $7,094
13  2007  $5,319
14  2006  $3,092
15  2005  $1,843

In [31]:
gme_revenue['Date'] = gme_revenue['Date'].astype(object)

In [32]:
gme_revenue.dtypes

Date       object
Revenue    object
dtype: object

In [33]:
gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"")

<ipython-input-33-f03406dd5959>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"")


In [34]:
gme_revenue.tail()

Date Revenue
11  2009    8806
12  2008    7094
13  2007    5319
14  2006    3092
15  2005    1843

In [35]:
gme_revenue['Date'] = gme_revenue['Date'].astype(int)

In [36]:
gme_revenue.dtypes

Date        int64
Revenue    object
dtype: object

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [37]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

ValueError: ignored

Figure({
    'data': [{'name': 'Share Price',
              'type': 'scatter',
              'x': array([datetime.datetime(2010, 6, 29, 0, 0),
                          datetime.datetime(2010, 6, 30, 0, 0),
                          datetime.datetime(2010, 7, 1, 0, 0), ...,
                          datetime.datetime(2021, 6, 10, 0, 0),
                          datetime.datetime(2021, 6, 11, 0, 0),
                          datetime.datetime(2021, 6, 14, 0, 0)], dtype=object),
              'xaxis': 'x',
              'y': array([  1.59266698,   1.58866704,   1.46399999, ..., 203.37333679,
                          203.29666138, 205.89666748]),
              'yaxis': 'y'},
             {'name': 'Revenue',
              'type': 'scatter',
              'x': array([datetime.datetime(2021, 1, 1, 0, 0),
                          datetime.datetime(2020, 1, 1, 0, 0),
                          datetime.datetime(2019, 1, 1, 0, 0),
                          datetime.datetime(2018, 1, 1, 0, 0),
                          datetime.datetime(2017, 1, 1, 0, 0),
                          datetime.datetime(2016, 1, 1, 0, 0),
                          datetime.datetime(2015, 1, 1, 0, 0),
                          datetime.datetime(2014, 1, 1, 0, 0),
                          datetime.datetime(2013, 1, 1, 0, 0),
                          datetime.datetime(2012, 1, 1, 0, 0),
                          datetime.datetime(2011, 1, 1, 0, 0),
                          datetime.datetime(2010, 1, 1, 0, 0),
                          datetime.datetime(2009, 1, 1, 0, 0)], dtype=object),
              'xaxis': 'x2',
              'y': array([53823., 31536., 24578., 21461., 11759.,  7000.,  4046.,  3198.,  2013.,
                            413.,   204.,   117.,   112.]),
              'yaxis': 'y2'}],
    'layout': {'annotations': [{'font': {'size': 16},
                                'showarrow': False,
                                'text': 'Historical Share Price',
                                'x': 0.5,
                                'xanchor': 'center',
                                'xref': 'paper',
                                'y': 0.9999999999999999,
                                'yanchor': 'bottom',
                                'yref': 'paper'},
                               {'font': {'size': 16},
                                'showarrow': False,
                                'text': 'Historical Revenue',
                                'x': 0.5,
                                'xanchor': 'center',
                                'xref': 'paper',
                                'y': 0.35,
                                'yanchor': 'bottom',
                                'yref': 'paper'}],
               'height': 900,
               'showlegend': False,
               'template': '...',
               'title': {'text': 'Tesla'},
               'xaxis': {'anchor': 'y',
                         'domain': [0.0, 1.0],
                         'matches': 'x2',
                         'rangeslider': {'visible': True},
                         'showticklabels': False,
                         'title': {'text': 'Date'}},
               'xaxis2': {'anchor': 'y2', 'domain': [0.0, 1.0], 'title': {'text': 'Date'}},
               'yaxis': {'anchor': 'x',
                         'domain': [0.6499999999999999, 0.9999999999999999],
                         'title': {'text': 'Price ($US)'}},
               'yaxis2': {'anchor': 'x2', 'domain': [0.0, 0.35], 'title': {'text': 'Revenue ($US Millions)'}}}
})

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [42]:
#I fixed the fuction because it didn't work properly when I used the webscraping method 'read_html' insted of 'BeatifulSoup'
#The problem occur due to a comparison between different data types.
#Specifically, we were attempting to compare a date string ('2021-04-30') with the 'Date' column in our DataFrame, which seems to contain datetime objects.
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing=.3)

    # Convert the date string to datetime objects
    stock_data_specific = stock_data[pd.to_datetime(stock_data['Date']) <= pd.to_datetime('2021-06-14')]
    revenue_data_specific = revenue_data[pd.to_datetime(revenue_data['Date']) <= pd.to_datetime('2021-04-30')]

    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)

    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)

    fig.update_layout(showlegend=False, height=900, title=stock, xaxis_rangeslider_visible=True)

    # Display the figure using IPython's display method
    display(fig)

In [43]:
make_graph(gme_data, gme_revenue, 'GameStop')

ValueError: ignored

Figure({
    'data': [{'name': 'Share Price',
              'type': 'scatter',
              'x': array([datetime.datetime(2002, 2, 13, 0, 0),
                          datetime.datetime(2002, 2, 14, 0, 0),
                          datetime.datetime(2002, 2, 15, 0, 0), ...,
                          datetime.datetime(2021, 6, 10, 0, 0),
                          datetime.datetime(2021, 6, 11, 0, 0),
                          datetime.datetime(2021, 6, 14, 0, 0)], dtype=object),
              'xaxis': 'x',
              'y': array([ 1.69166613,  1.68325055,  1.67483389, ..., 55.09749985, 58.33499908,
                          57.36000061]),
              'yaxis': 'y'},
             {'name': 'Revenue',
              'type': 'scatter',
              'x': array([datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2),
                          datetime.datetime(1970, 1, 1, 0, 0, 0, 2)], dtype=object),
              'xaxis': 'x2',
              'y': array([6466., 8285., 8547., 7965., 9364., 9296., 9040., 8887., 9551., 9474.,
                          9078., 8806., 7094., 5319., 3092., 1843.]),
              'yaxis': 'y2'}],
    'layout': {'annotations': [{'font': {'size': 16},
                                'showarrow': False,
                                'text': 'Historical Share Price',
                                'x': 0.5,
                                'xanchor': 'center',
                                'xref': 'paper',
                                'y': 0.9999999999999999,
                                'yanchor': 'bottom',
                                'yref': 'paper'},
                               {'font': {'size': 16},
                                'showarrow': False,
                                'text': 'Historical Revenue',
                                'x': 0.5,
                                'xanchor': 'center',
                                'xref': 'paper',
                                'y': 0.35,
                                'yanchor': 'bottom',
                                'yref': 'paper'}],
               'height': 900,
               'showlegend': False,
               'template': '...',
               'title': {'text': 'GameStop'},
               'xaxis': {'anchor': 'y',
                         'domain': [0.0, 1.0],
                         'matches': 'x2',
                         'rangeslider': {'visible': True},
                         'showticklabels': False,
                         'title': {'text': 'Date'}},
               'xaxis2': {'anchor': 'y2', 'domain': [0.0, 1.0], 'title': {'text': 'Date'}},
               'yaxis': {'anchor': 'x',
                         'domain': [0.6499999999999999, 0.9999999999999999],
                         'title': {'text': 'Price ($US)'}},
               'yaxis2': {'anchor': 'x2', 'domain': [0.0, 0.35], 'title': {'text': 'Revenue ($US Millions)'}}}
})

<h2>About the Authors:</h2>

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
